[mlir] Add missing fields in DICompositeTypeAttr.

abidh opened the pull request:
The fortran arrays use 'dataLocation', 'rank', 'allocated' and 'associated' fields of the DICompositeType. These were not available in 'DICompositeTypeAttr'. This PR adds the missing fields.

abidh committed:
[mlir] Add missing fields in DICompositeTypeAttr.

The fortran arrays use 'dataLocation', 'rank', 'allocated' and
'associated' fields of the DICompositeType. These were not available
in 'DICompositeTypeAttr'. This PR adds the missing fields.

gysit commented on mlir/include/mlir-c/Dialect/LLVM.h:
```diff
@@ -239,7 +239,9 @@ MLIR_CAPI_EXPORTED MlirAttribute mlirLLVMDICompositeTypeAttrGet(
     MlirAttribute baseType, int64_t flags, uint64_t sizeInBits,
-    uint64_t alignInBits, intptr_t nElements, MlirAttribute const *elements);
+    uint64_t alignInBits, intptr_t nElements, MlirAttribute const *elements,
+    MlirAttribute location, MlirAttribute rank, MlirAttribute allocated,
```
```suggestion
    MlirAttribute dataLocation, MlirAttribute rank, MlirAttribute allocated,
```
nit: should it be dataLocation as well for consistency?

gysit commented on mlir/lib/CAPI/Dialect/LLVM.cpp:
```diff
@@ -163,7 +163,9 @@ MlirAttribute mlirLLVMDICompositeTypeAttrGet(
     MlirAttribute baseType, int64_t flags, uint64_t sizeInBits,
-    uint64_t alignInBits, intptr_t nElements, MlirAttribute const *elements) {
+    uint64_t alignInBits, intptr_t nElements, MlirAttribute const *elements,
+    MlirAttribute location, MlirAttribute rank, MlirAttribute allocated,
```
```suggestion
    MlirAttribute dataLocation, MlirAttribute rank, MlirAttribute allocated,
```

gysit commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -484,3 +484,39 @@ llvm.mlir.global @global_variable() {dbg_expr = #di_global_variable_expression}
 // CHECK: ![[SUBROUTINE]] = !DISubroutineType(types: ![[SR_TYPES:[0-9]+]])
 // CHECK: ![[SR_TYPES]] = !{![[COMP]]}
+
+// -----
```
```suggestion
// -----

```
nit: missing newline

gysit commented on mlir/test/Target/LLVMIR/Import/composite-type-fortran.ll:
```diff
@@ -0,0 +1,27 @@
+; CHECK-SAME: rank = <[DW_OP_push_object_address, DW_OP_plus_uconst(16), DW_OP_deref]>
+; CHECK-SAME: allocated = <[DW_OP_lit0, DW_OP_ne]>
+; CHECK-SAME: associated = <[DW_OP_lit0, DW_OP_eq]>
+
+
```
```suggestion
```
nit: drop trailing newlines.

gysit commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -484,3 +484,39 @@ llvm.mlir.global @global_variable() {dbg_expr = #di_global_variable_expression}
+  compileUnit = #cu, scope = #cu, name = "fn_with_composite", file = #file,
+  subprogramFlags = "Definition|Optimized", type = #spType0
+>
+llvm.func @fn_with_composite() {
+  llvm.return
+}loc(fused<#sp0>["foo.mlir":1:1])
+// CHECK: !DISubprogram(name: "fn_with_composite"{{.*}}type: ![[TY1:[0-9]+]]{{.*}})
+// CHECK: [[TY1]] = !DISubroutineType({{.*}}types: ![[TYPES:[0-9]+]])
+// CHECK: ![[TYPES]] = !{null, ![[COMPTY:[0-9]+]]}
+// CHECK: ![[COMPTY]] = !DICompositeType
+// CHECK-SAME: dataLocation: !DIExpression(DW_OP_push_object_address, DW_OP_deref)
+// CHECK-SAME: associated: !DIExpression(DW_OP_lit0, DW_OP_eq)
+// CHECK-SAME: allocated: !DIExpression(DW_OP_lit0, DW_OP_ne)
+// CHECK-SAME: rank: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 16, DW_OP_deref)
```
```suggestion

llvm.func @fn_with_composite() {
  llvm.return
} loc(fused<#sp0>["foo.mlir":1:1])

// CHECK: !DICompositeType(
// CHECK-SAME: dataLocation: !DIExpression(DW_OP_push_object_address, DW_OP_deref)
// CHECK-SAME: associated: !DIExpression(DW_OP_lit0, DW_OP_eq)
// CHECK-SAME: allocated: !DIExpression(DW_OP_lit0, DW_OP_ne)
// CHECK-SAME: rank: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 16, DW_OP_deref)
```

gysit replied:
nit: I would check for the composite type itself here but not necessarily for the enclosing subroutine type. Also added some whitespaces to format the function.

abidh replied:
I used this approach initially but observed that test will fail because `FileCheck` will match an earlier `!DICompositeType` in the file. In the updated patch, I used your suggestion but added a `CHECK-LABEL` to avoid the wrong match.

gysit reviewed (commented):
Thanks!

Looks good. I added some minor formatting / naming comments.

abidh committed:
Update mlir/include/mlir-c/Dialect/LLVM.h

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Update mlir/test/Target/LLVMIR/Import/composite-type-fortran.ll

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Update mlir/test/Target/LLVMIR/llvmir-debug.mlir

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Update mlir/lib/CAPI/Dialect/LLVM.cpp

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Handle review comments

abidh committed:
Improve test case.

gysit commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -484,3 +484,38 @@ llvm.mlir.global @global_variable() {dbg_expr = #di_global_variable_expression}
+#sp0 = #llvm.di_subprogram<
+  compileUnit = #cu, scope = #cu, name = "fn_with_composite", file = #file,
+  subprogramFlags = "Definition|Optimized", type = #spType0
+>
+llvm.func @fn_with_composite() {
+  llvm.return
+}loc(fused<#sp0>["foo.mlir":1:1])
+// CHECK-LABEL: define void @fn_with_composite()
```
```suggestion
>

llvm.func @fn_with_composite() {
  llvm.return
} loc(fused<#sp0>["foo.mlir":1:1])

// CHECK-LABEL: define void @fn_with_composite()
```
ultra nit: Added some more spaces here. The CHECK-LABEL makes a lot of sense!

gysit reviewed (approved):
Thanks LGTM modulo last nit!

abidh committed:
Fix flang build.

The #92568 introduced call of DICompositeTypeAttr::get in flang.
This commit adjusts the call to fix the build.

I initially opened the #93516 for it but was advised to add this
change in this PR.

abidh commented:
The #92568 introduced call of [DICompositeTypeAttr::get](https://github.com/llvm/llvm-project/blob/1383cb6e648ba9f4516fa2317f948116c4f35c10/flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp#L72) in flang. That will need to be updated once this PR is merged. I have opened #93516 for that.

gysit commented:
There seem to be issues with the build bots still. Once they are fixed this commit can be landed!

abidh commented:
> The #92568 introduced call of [DICompositeTypeAttr::get](https://github.com/llvm/llvm-project/blob/1383cb6e648ba9f4516fa2317f948116c4f35c10/flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp#L72) in flang

After opening of this PR, #92568 went in which added a call to DICompositeTypeAttr::get in flang that was causing the build failure. I have pushed a commit to adjust the call. I will wait for bots to turn green before landing.
